compiler/qsc_circuit/src/operations.rs

```diff
@@ -0,0 +1,296 @@
+/// need be allocated to run this operation for the purposes of circuit generation.
+#[must_use]
+pub fn qubit_param_info(item: &Item) -> Option<(Vec<u32>, u32)> {
+    if let ItemKind::Callable(decl) = &item.kind {
+        let (qubit_param_dimensions, total_num_qubits) = get_qubit_param_info(&decl.input.ty);
+
+        if !qubit_param_dimensions.is_empty() {
+            return Some((qubit_param_dimensions, total_num_qubits));
+        }
+    }
+    None
+}
+
+/// Returns an entry expression to directly invoke the operation
+/// for the purposes of generating a circuit for it.
+///
+/// `operation_expr` is the source for the expression that refers to the operation,
+/// e.g. "Test.Foo" or "qs => H(qs[0])".
+///
+/// If the item is not a callable, returns `None`.
+/// If the callable takes any non-qubit parameters, returns `None`.
+#[must_use]
+pub fn entry_expr_for_qubit_operation(item: &Item, operation_expr: &str) -> Option<String> {
+    if let Some((qubit_param_dimensions, total_num_qubits)) = qubit_param_info(item) {
+        return Some(operation_circuit_entry_expr(
+            operation_expr,
+            &qubit_param_dimensions,
+            total_num_qubits,
+        ));
+    }
+    None
+}
+
+/// Generates the entry expression to call the operation described by `params`.
+/// The expression allocates qubits and invokes the operation.
+#[must_use]
+fn operation_circuit_entry_expr(
+    operation_expr: &str,
+    qubit_param_dimensions: &[u32],
+    total_num_qubits: u32,
+) -> String {
+    let alloc_qubits = format!("use qs = Qubit[{total_num_qubits}];");
+
+    let mut qs_start = 0;
+    let mut call_args = vec![];
+    for dim in qubit_param_dimensions {
+        let dim = *dim;
+        let qs_len = NUM_QUBITS.pow(dim);
+        // Q# ranges are end-inclusive
+        let qs_end = qs_start + qs_len - 1;
+        if dim == 0 {
+            call_args.push(format!("qs[{qs_start}]"));
+        } else {
+            // Array argument - use a range to index
+            let mut call_arg = format!("qs[{qs_start}..{qs_end}]");
+            for _ in 1..dim {
+                // Chunk the array for multi-dimensional array arguments
+                call_arg = format!("Microsoft.Quantum.Arrays.Chunks({NUM_QUBITS}, {call_arg})");
+            }
+            call_args.push(call_arg);
+        }
+        qs_start = qs_end + 1;
+    }
+
+    let call_args = call_args.join(", ");
+
+    // We don't reset the qubits since we don't want reset gates
+    // included in circuit output.
+    // We also don't measure the qubits but we have to return a result
+    // array to satisfy Base Profile.
+    format!(
+        r#"{{
+            {alloc_qubits}
+            ({operation_expr})({call_args});
+            let r: Result[] = [];
+            r
+        }}"#
+    )
+}
+
+/// The number of qubits to allocate for each qubit array
+/// in the operation arguments.
+static NUM_QUBITS: u32 = 2;
```

Comment:
Would using `const` be more appropriate than using `static` in this case?

Reply:
I forgot `const` was a thing! Updating.